rs/embedders/src/wasmtime_embedder.rs

```diff
@@ -332,6 +336,42 @@ impl WasmtimeEmbedder {
         self.pre_instantiate(&module)
     }
 
+    /// TODO(EXC-1800): Replace this with `wasmtime::Module::deserialize_open_file`.
+    fn deserialize_from_file(&self, serialized_module: &File) -> HypervisorResult<Module> {
+        let mmap_size = serialized_module.metadata().unwrap().size() as usize;
+        let mmap_ptr = unsafe {
+            mmap(
+                std::ptr::null_mut(),
+                mmap_size,
+                ProtFlags::PROT_READ,
+                MapFlags::MAP_PRIVATE,
+                serialized_module.as_raw_fd(),
+                0,
+            )
+        }
+        .unwrap_or_else(|err| panic!("Module deserialization failed: {:?}", err))
+            as *mut u8;
+        let bytes = unsafe { std::slice::from_raw_parts(mmap_ptr, mmap_size) };
+        unsafe {
+            Module::deserialize(&self.create_engine()?, bytes).map_err(|err| {
+                HypervisorError::WasmEngineError(WasmEngineError::FailedToDeserializeModule(
+                    format!("{:?}", err),
+                ))
+            })
+        }
+    }
+
+    pub fn read_file_and_pre_instantiate(
+        &self,
+        serialized_module: File,
+    ) -> HypervisorResult<InstancePre<StoreData>> {
+        // TODO(EXC-1800): Switch to new wasmtime API and remove leaking the
+        // file.
+        let module = self.deserialize_from_file(&serialized_module)?;
+        let _ = serialized_module.into_raw_fd();
```

Comment:
This drops the OS fd, right? Making it possible for the OS to remove the file from disk.

Reply:
Or does the mmapped `module` still prevent that?

Reply:
I realized that `RawFd` is an alias for `c_int`... nevermind.

Reply:
This actually leaks the file descriptor (see https://doc.rust-lang.org/std/os/fd/trait.IntoRawFd.html#tymethod.into_raw_fd) because we never call `close` on the resulting integer. When we switch to the `wasmtime` call, a reference to the fd will be stored in the resulting `wasmtime::Module` object so that it's properly closed when the `Module` is dropped.

Reply:
So we temporarily leak it, but this int ends up in another rust-managed fd? How, since it is dropped here? How do we send it to the receiver?

Reply:
I mean with this current version of the code we permanently leak it in the sandbox, so it'll only get closed when the sandbox exits. But in a later PR that addresses EXC-1800, we'll switch from using `self.deserialize_from_file` to [`wasmtime::Module::deserialize_open_file`](https://docs.wasmtime.dev/api/wasmtime/struct.Module.html#method.deserialize_open_file) which will take ownership of the `File` and store it in the resulting `Module`. Then we can also remove this line that leaks it.

Reply:
Thanks for clarifying!